Add BlobContainer.writeBlobAtomic()

tlrx opened the pull request:
This commit adds a new writeBlobAtomic() method to the BlobContainer interface that can be implemented by repository implementations which support atomic writes operations.

When the repository does not support atomic writes, this method just delegate the write operation to the usual writeBlob() method.

As a follow up method, I think we could just remove the `BlobContainer.move()` method now it's just used during repository verification.

Related to #30680

elasticmachine commented:
Pinging @elastic/es-distributed

ywelsch commented on server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:
```diff
@@ -131,6 +132,28 @@ public void writeBlob(String blobName, InputStream inputStream, long blobSize) t
 
+    @Override
+    public void writeBlobAtomic(final String blobName, final InputStream inputStream, final long blobSize) throws IOException {
+        final Path tempBlobPath = path.resolve("pending-" + blobName + "-" + UUIDs.randomBase64UUID());
```
There is a relation between the name here and between `ChecksumBlobStoreFormat.isTempBlobName`, which is now easy to break. I think we should make this more explicit.

tlrx replied:
I agree

tlrx commented:
> I would prefer if you also replace the last occurrence of move, and then remove the move method.

I removed the last occurance (but added one in test). I'll remove the method in a follow up pull request if that's fine for you (I expect to remove many things in repository plugins related to this).



> I wonder if we need to inject more failures now so that FSBlobContainer

I agree. I updated the code, can you have another look please? 

I also move two tests classes were they should be and fix the checkstyle violations on the way.

ywelsch commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java:
```diff
@@ -74,6 +74,28 @@
 
+    /**
+     * Reads blob content from the input stream and writes it to the container in a new blob with the given name,
+     * using an atomic write operation if then implementation supports it. When atomic writes are not supported,
```
s/then/the/
Also it's rather the opposite. If the implementation support atomic writes by default, it then delegates to writeBlob...?

tlrx replied:
"When the BlobContainer implementation does not provide a specific implementation for wireBlobAtomic, then the usual writeBlob method is used." Would that be better?

ywelsch replied:
yes

ywelsch commented on server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:
```diff
@@ -131,6 +133,41 @@ public void writeBlob(String blobName, InputStream inputStream, long blobSize) t
+            }
+            Files.move(tempBlobPath, blobPath, StandardCopyOption.ATOMIC_MOVE);
+        } finally {
+            IOUtils.deleteFilesIgnoringExceptions(tempBlobPath);
```
Should we really ignore all exceptions here? This means we can up with lingering files without any warnings.
I preferred the previous approach of:

```
        } catch (IOException ex) {		
            // temporary blob creation or move failed - try cleaning up		
            try {		
                snapshotsBlobContainer.deleteBlobIgnoringIfNotExists(tempBlobName);		
            } catch (IOException e) {		
                ex.addSuppressed(e);		
            }		
            throw ex;		
          }
```

tlrx replied:
Right, good point

tlrx commented:
Thanks @ywelsch  - I applied your feedback. Can you have another look please?

ywelsch commented on server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:
```diff
@@ -131,6 +133,49 @@ public void writeBlob(String blobName, InputStream inputStream, long blobSize) t
+            }
+            throw ex;
+        } finally {
+            IOUtils.deleteFilesIgnoringExceptions(tempBlobPath);
```
this line should be removed now.

tlrx replied:
raaah thanks

I should pay a :beer: every time I forgot something like that

tlrx commented:
@elasticmachine test this please

tlrx commented:
Thanks @ywelsch
